### PR TITLE
Impl set datapoints

### PR DESCRIPTION
**Pull request by BjoernAtBosch**
## Describe your changes

C++ SDK is extended to support setting multiple data points in a single atomic step.

## Issue ticket number and link

AB#11773

## Checklist - Manual tasks

* [x] Examples are executing successfully
* [x] Created/updated unit tests. Code Coverage percentage on new code shall be >= 80%.
* [ ] Created/updated integration tests.
* [x] Devcontainer can be opened successfully
* [x] Devcontainer can be opened successfully behind a corporate proxy
* [x] Devcontainer can be re-built successfully
* [ ] Extended the documentation (e.g. README.md, CONTRIBUTING.md, Velocitas)

### CODE REVIEW

File: examples/set-data-points/AppManifest.json [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                    "version": "v0.17.0"
+                },
+                {
+                    "name": "feedercan",
+                    "image": "ghcr.io/eclipse/kuksa.val.feeders/dbc2val",
+                    "version": "v0.1.0"
+                }
+            ],
+            "python": {
+                "version": "3.9"
```
Comment (MP91) [inline examples/set-data-points/AppManifest.json:33]: shouldn't this be the same values as in the seatadjuster-app?

File: AppManifest.json [hunk 2]
```diff
@@ -1,14 +1,15 @@
 [
     {
-        "Name": "seatadjuster",
-        "Port": 50008,
+        "Name": "sampleapp",
+        "Port": -1,
```
Comment (MP91) [inline AppManifest.json:4]: is this intended? couldn't we remove the line completely?

File: sdk/include/sdk/DataPointValues.h [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    // Unexpected internal error
+    INTERNAL_ERROR,
+};
+
+std::string toString(DataPointFailure);
+
+class DataPointValue {
+public:
+    enum class Type {
+        INVALID,
```
Comment (MP91) [inline sdk/include/sdk/DataPointValues.h:68]: why not 
```suggestion
        FAILURE,
```
?

File: sdk/include/sdk/DataPointValues.h [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+};
+
+using DataPointMap_t = std::map<std::string, std::shared_ptr<DataPointValue>>;
+
+/**
+ * @brief Result of an operation which returns multiple data points.
+ *        Provides typed access to obtained data points.
+ *
+ */
+class DataPointValues final {
```
Comment (MP91) [inline sdk/include/sdk/DataPointValues.h:208]: What about
```suggestion
class DataPointValueMap final {
```
I think just adding a `s` will lead to confusions.

**Review by MP91 — COMMENTED** [review]
I just did a qick review and added some comments. I would like to wait for @doosuu  for a thorough review

File: sdk/include/sdk/DataPointValues.h [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    DataPointFailure m_failure{DataPointFailure::NONE};
+};
+
+template <typename T> DataPointValue::Type getValueType() {
+    static_assert(std::is_same<T, std::false_type>::value, "Value type not supported!");
+    return DataPointValue::Type::INVALID;
+}
+
+template <> inline DataPointValue::Type getValueType<bool>() { return DataPointValue::Type::BOOL; }
+template <> inline DataPointValue::Type getValueType<std::vector<bool>>() {
```
Comment (doosuu) [inline sdk/include/sdk/DataPointValues.h:124]: Maybe we should typedef the std::vector<T> to something to be able to replace vector with a different implementation easily.

File: sdk/include/sdk/QueryBuilder.h [hunk 6]
```diff
@@ -17,44 +17,74 @@
...
 /**
- * @brief Builder for where subclauses within the VDB query.
+ * @brief Builder for VDB queries.
  *
  */
-class WhereClauseBuilder final {
+class QueryBuilder final {
 public:
-    WhereClauseBuilder() = default;
+    static QueryBuilder select(const DataPoint& dataPoint);
```
Comment (doosuu) [inline sdk/include/sdk/QueryBuilder.h:35]: Missing doxygen for all methods in QueryBuilder and WhereClauseBuilder

File: sdk/include/sdk/VehicleApp.h [hunk 7]
```diff
@@ -111,20 +111,21 @@ class VehicleApp {
...
             });
     }
 
     /**
-     * @brief Subscribes to the query string for data points.
+     * @brief Subscribes to the query for data points.
      *
-     * @param queryString   The query string to subscribe to.
-     * @return AsyncSubscriptionPtr_t<DataPointsResult>   The subscription to the data points.
+     * @param query   The query to subscribe to.
```
Comment (doosuu) [inline sdk/include/sdk/VehicleApp.h:125]: Parameter is still called queryString

File: sdk/include/sdk/vdb/IDataPointValueProvider.h [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef VEHICLE_APP_SDK_DATAPOINTVALUEPROVIDER_H
```
Comment (doosuu) [inline sdk/include/sdk/vdb/IDataPointValueProvider.h:17]:
```suggestion
#ifndef VEHICLE_APP_SDK_IDATAPOINTVALUEPROVIDER_H
```

File: sdk/src/sdk/DataPointValues.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        return "DataPointFailure::UNKNOWN_DATAPOINT";
+    case DataPointFailure::ACCESS_DENIED:
+        return "DataPointFailure::ACCESS_DENIED";
+    case DataPointFailure::INTERNAL_ERROR:
+        return "DataPointFailure::INTERNAL_ERROR";
+    default:
+        logger().error(
+            "velocitas::toString(DataPointFailure): Unknown DataPointFailure enum value {}",
+            static_cast<unsigned int>(failure));
+        assert(false);
```
Comment (doosuu) [inline sdk/src/sdk/DataPointValues.cpp:46]: Do we need the assert here?

File: sdk/src/sdk/grpc/GrpcDataPointValueProvider.cpp [hunk 10]
```diff
@@ -27,58 +29,62 @@ const sdv::databroker::v1::Datapoint& GrpcDataPointValueProvider::getDataPoint()
...
+        return DataPointFailure::NOT_AVAILABLE;
+    case sdv::databroker::v1::Datapoint_Failure_UNKNOWN_DATAPOINT:
+        return DataPointFailure::UNKNOWN_DATAPOINT;
+    case sdv::databroker::v1::Datapoint_Failure_ACCESS_DENIED:
+        return DataPointFailure::ACCESS_DENIED;
+    case sdv::databroker::v1::Datapoint_Failure_INTERNAL_ERROR:
+        return DataPointFailure::INTERNAL_ERROR;
+    default:
+        logger().error("Unknown 'DataPointFailure': {}", m_datapoint.failure_value());
+        assert(false);
```
Comment (doosuu) [inline sdk/src/sdk/grpc/GrpcDataPointValueProvider.cpp:46]: Assert needed?

File: sdk/src/sdk/grpc/VehicleDataBrokerClient.cpp [hunk 11]
```diff
@@ -60,56 +61,197 @@ std::string VehicleDataBrokerClient::getVdbEndpointAddress() {
...
+        return sdv::databroker::v1::Datapoint_Failure_NOT_AVAILABLE;
+    case DataPointFailure::UNKNOWN_DATAPOINT:
+        return sdv::databroker::v1::Datapoint_Failure_UNKNOWN_DATAPOINT;
+    case DataPointFailure::ACCESS_DENIED:
+        return sdv::databroker::v1::Datapoint_Failure_ACCESS_DENIED;
+    case DataPointFailure::INTERNAL_ERROR:
+        return sdv::databroker::v1::Datapoint_Failure_INTERNAL_ERROR;
+    default:
+        logger().error("Unknown 'DataPointFailure': {}", static_cast<unsigned int>(failure));
+        assert(false);
```
Comment (doosuu) [inline sdk/src/sdk/grpc/VehicleDataBrokerClient.cpp:79]: Assert needed?

File: sdk/src/sdk/grpc/VehicleDataBrokerClient.cpp [hunk 12]
```diff
@@ -60,56 +61,197 @@ std::string VehicleDataBrokerClient::getVdbEndpointAddress() {
     return fmt::format("localhost:{}", daprGrpcPort);
 }
 
-std::shared_ptr<DataPoint>
+static sdv::databroker::v1::Datapoint_Failure
+dataPointFailure_to_grpcFailure(DataPointFailure failure) {
```
Comment (doosuu) [inline sdk/src/sdk/grpc/VehicleDataBrokerClient.cpp:65]: Invalid function name. Should be in camelCase

File: sdk/tests/unit/CMakeLists.txt [hunk 13]
```diff
@@ -20,10 +23,13 @@ add_dependencies(${TARGET_NAME}
...
+    gmock
 )
 
 include(GoogleTest)
 
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../mocks)
+
 if(NOT CMAKE_TOOLCHAIN_FILE)
     gtest_discover_tests(${TARGET_NAME})
 endif()
```
Comment (doosuu) [inline sdk/tests/unit/CMakeLists.txt:35]: Add missing trailing newline

**Review by doosuu — CHANGES_REQUESTED** [review]

**Review by emsameen — COMMENTED** [review]
I had a quick look into the structure and I add few comments. 
I also observed the folder name of the example app is called **datapoint-set** instead of **set-data-points** this is something you may want to update.

File: sdk/src/sdk/DataPointValues.cpp (hunk 9 above)
Comment (BjoernAtBosch) [reply sdk/src/sdk/DataPointValues.cpp:46]: It's about "forcing" the developer to regard newly added enum values here.
Asserts are only drawn in debug build, not release.

File: sdk/include/sdk/DataPointValues.h (hunk 5 above)
Comment (BjoernAtBosch) [reply sdk/include/sdk/DataPointValues.h:124]: I'd prefer doings this in a later refactoring

File: examples/seat-adjuster/src/SeatAdjusterApp.cpp [hunk 14]
```diff
@@ -70,7 +70,7 @@ void SeatAdjusterApp::onStart() {
         ->onError([this](auto&& status) { onErrorTopic(std::forward<decltype(status)>(status)); });
 }
 
-void SeatAdjusterApp::onSpeedChanged(const velocitas::DataPointsResult& dataPoints) {
+void SeatAdjusterApp::onSpeedChanged(const velocitas::DataPointValues& dataPoints) {
```
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply examples/seat-adjuster/src/SeatAdjusterApp.cpp:73]: Maybe in this special context ~Result would better fit. But the renaming is about the type of the result/response.
For my understanding we have "data points" (which are similar to "variables" in program code) and a data point (i.e. a variable) has a certain "value". I not sure, what the "result" of a data point should be? Of course, there can be the "result" of querying a data point's current state - which then is a "value" ...
I'm not sure what's your point with the bull point, here?

File: sdk/include/sdk/DataPointValues.h (hunk 4 above)
Comment (BjoernAtBosch) [reply sdk/include/sdk/DataPointValues.h:208]: Class is just used in query (get/subscribe) case. Renaming to DataPointReply to be in sync with Python implementation

File: examples/seat-adjuster/src/SeatAdjusterApp.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply examples/seat-adjuster/src/SeatAdjusterApp.cpp:73]: Class is just used in query (get/subscribe) case. Renaming to DataPointReply to be in sync with Python implementation

File: sdk/src/sdk/DataPointValues.cpp (hunk 9 above)
Comment (BjoernAtBosch) [reply sdk/src/sdk/DataPointValues.cpp]: Will add TechDebt or task for fixing this

File: AppManifest.json [hunk 15]
```diff
@@ -23,6 +23,11 @@
                     "image": "ghcr.io/boschglobal/kuksa.val/databroker",
                     "version": "0.0.2"
                 },
+                {
```
Comment (doosuu) [inline AppManifest.json:26]: This should not be in the AppManifest.json. Why was this added?

File: examples/seat-adjuster/AppManifest.json [hunk 16]
```diff
@@ -20,18 +20,20 @@
                 },
                 {
                     "name": "databroker",
-                    "image": "ghcr.io/eclipse/kuksa.val/databroker",
-                    "version": "v0.17.0"
+                    "image": "ghcr.io/boschglobal/kuksa.val/databroker",
+                    "version": "0.0.2"
+                },
+                {
+                    "name": "databroker_cli",
```
Comment (doosuu) [inline examples/seat-adjuster/AppManifest.json:27]: This should not be in the AppManifest.json. Why was this added?

File: examples/set-data-points/AppManifest.json [hunk 17]
```diff
@@ -20,21 +20,23 @@
                 },
                 {
                     "name": "databroker",
-                    "image": "ghcr.io/eclipse/kuksa.val/databroker",
-                    "version": "v0.17.0"
+                    "image": "ghcr.io/boschglobal/kuksa.val/databroker",
+                    "version": "0.0.2"
+                },
+                {
+                    "name": "databroker_cli",
```
Comment (doosuu) [inline examples/set-data-points/AppManifest.json:27]: This should not be in the AppManifest.json. Why was this added?

File: sdk/include/sdk/DataPointReply.h [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+ */
+class DataPointReply final {
+public:
+    DataPointReply() = default;
+
+    DataPointReply(DataPointMap_t&& dataPointsMap)
+        : m_dataPointsMap(std::move(dataPointsMap)) {}
+
+    /**
+     * @brief Get the desired data point from the result.
```
Comment (doosuu) [inline sdk/include/sdk/DataPointReply.h:50]:
```suggestion
     * @brief Get the desired data point from the reply.
```

File: sdk/include/sdk/DataPointReply.h [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+     *
+     * @tparam TDataPointType   The type of the data point to return.
+     * @param dataPoint         The data point to query from the result.
+     * @return std::shared_ptr<TDataPointType>  The data point contained in the result.
```
Comment (doosuu) [inline sdk/include/sdk/DataPointReply.h:54]:
```suggestion
     * @return std::shared_ptr<TDataPointType>  The data point contained in the reply.
```

File: sdk/include/sdk/DataPointReply.h [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+            return std::dynamic_pointer_cast<
+                TypedDataPointValue<typename TDataPointType::value_type>>(result);
+        }
+
+        return std::make_shared<TypedDataPointValue<typename TDataPointType::value_type>>(
+            result->getPath(), result->getFailure(), result->getTimestamp());
+    }
+
+    /**
+     * @brief Check if the result is empty.
```
Comment (doosuu) [inline sdk/include/sdk/DataPointReply.h:77]:
```suggestion
     * @brief Check if the reply is empty.
```

File: sdk/include/sdk/QueryBuilder.h [hunk 21]
```diff
@@ -52,39 +76,68 @@ class QueryBuilder final {
...
-        : m_parent{parent} {
-        m_parent->m_queryContext.emplace_back("WHERE");
-        m_parent->m_queryContext.emplace_back(dataPoint.getPath());
-    }
-
+    /**
+     * @brief Adds a greater than condition for the chosen data point
+     *
+     * @param value   The value of the data point must be greater than this
+     * @return A reference to a WhereClauseBuilder for adding more conditions or triggering the
```
Comment (doosuu) [inline sdk/include/sdk/QueryBuilder.h:88]:
```suggestion
     * @return This instance for method chaining.
```

File: sdk/include/sdk/QueryBuilder.h [hunk 22]
```diff
@@ -52,39 +76,68 @@ class QueryBuilder final {
...
         m_parent->m_queryContext.emplace_back(">");
         m_parent->m_queryContext.emplace_back(std::to_string(value));
         return *this;
     }
 
+    /**
+     * @brief Adds a lesser than condition for the chosen data point
+     *
+     * @param value   The value of the data point must be smaller than this
+     * @return A reference to a WhereClauseBuilder for adding more conditions or triggering the
```
Comment (doosuu) [inline sdk/include/sdk/QueryBuilder.h:101]: See previous finding.

File: sdk/include/sdk/QueryBuilder.h [hunk 23]
```diff
@@ -52,39 +76,68 @@ class QueryBuilder final {
...
         m_parent->m_queryContext.emplace_back("<");
         m_parent->m_queryContext.emplace_back(std::to_string(value));
         return *this;
     }
 
+    /**
+     * @brief Adds a equals to condition for the chosen data point
+     *
+     * @param value   The value of the data point must be equal to this
+     * @return A reference to a WhereClauseBuilder for adding more conditions or triggering the
```
Comment (doosuu) [inline sdk/include/sdk/QueryBuilder.h:114]: See previous finding.

File: sdk/include/sdk/VehicleApp.h [hunk 24]
```diff
@@ -97,10 +97,9 @@ class VehicleApp {
      * @brief Get values for all provided data points from the data broker.
      *
      * @param dataPoints    Vector of data points to obtain values for.
-     * @return AsyncResultPtr_t<DataPointValues>  The result containing the data point
-     *     results for all requested data points.
+     * @return The result contains the data point values for all requested data points.
```
Comment (doosuu) [inline sdk/include/sdk/VehicleApp.h:100]:
```suggestion
     * @return The result containing/which contains the data point values for all requested data points.
```

File: sdk/include/sdk/grpc/GrpcDataPointValueProvider.h [hunk 25]
```diff
@@ -27,7 +27,7 @@ class GrpcDataPointValueProvider : public IDataPointValueProvider {
 public:
     explicit GrpcDataPointValueProvider(sdv::databroker::v1::Datapoint datapoint);
 
-    DataPointFailure         getDataPointFailure() const override;
+    DataPointValue::Failure  getDataPointValueFailure() const override;
```
Comment (doosuu) [inline sdk/include/sdk/grpc/GrpcDataPointValueProvider.h:30]: Simply call this `getFailure`? It already is a Value Provider for a single datapoint, hence the name can be short and concise.

File: sdk/include/sdk/vdb/IDataPointValueProvider.h [hunk 26]
```diff
@@ -40,7 +40,7 @@ class IDataPointValueProvider {
     IDataPointValueProvider& operator=(const IDataPointValueProvider&) = delete;
     IDataPointValueProvider& operator=(IDataPointValueProvider&&)      = delete;
 
-    [[nodiscard]] virtual DataPointFailure getDataPointFailure() const = 0;
+    [[nodiscard]] virtual DataPointValue::Failure getDataPointValueFailure() const = 0;
```
Comment (doosuu) [inline sdk/include/sdk/vdb/IDataPointValueProvider.h:43]: Simply call this `getFailure`? It already is a Value Provider for a single datapoint, hence the name can be short and concise.

**Review by doosuu — CHANGES_REQUESTED** [review]

File: .vscode/scripts/runtime/local/run-vehicledatabroker-cli.sh [hunk 27]
```diff
@@ -25,29 +25,23 @@ source $UTILS_DIRECTORY/get-appmanifest-data.sh
```
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply .vscode/scripts/runtime/local/run-vehicledatabroker-cli.sh:25]: Changed because it was referencing the wrong repo and was using GitHub API

File: AppManifest.json (hunk 15 above)
Comment (BjoernAtBosch) [reply AppManifest.json:26]: Removed as discussed

File: examples/set-data-points/AppManifest.json (hunk 17 above)
Comment (BjoernAtBosch) [reply examples/set-data-points/AppManifest.json:27]: Removed as discussed

**Review by doosuu — APPROVED** [review]

**Comment by doosuu** [comment]
LGTM

**Review by MP91 — APPROVED** [review]
